### PR TITLE
Add locking property

### PR DESCRIPTION
**Pull request by treysp**
- Add new locking property
- Create new property location POST_ALIAS
- Rename property location PRE_SCHEMA to POST_NAME

Teradata dialect:
- Allow REPLACE VIEW
- Allow SEL as abbreviation for SELECT
- Map BYTEINT to TokenType.SMALLINT

### CODE REVIEW

File: sqlglot/dialects/teradata.py [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
 from sqlglot.dialects.dialect import Dialect
 from sqlglot.tokens import TokenType
 
 
 class Teradata(Dialect):
+    class Tokenizer(tokens.Tokenizer):
+        KEYWORDS = {
+            **tokens.Tokenizer.KEYWORDS,
+            "BYTEINT": TokenType.SMALLINT,
+            "SEL": TokenType.SELECT,
```
Comment (tobymao) [inline sqlglot/dialects/teradata.py:13]: lol wat

File: sqlglot/expressions.py [hunk 2]
```diff
@@ -1479,12 +1489,25 @@ class Properties(Expression):
...
+    #     table a [POST_NAME]
+    #     as [POST_ALIAS] (select * from b)
+    #     index (c) [POST_INDEX]
     class Location(AutoName):
         POST_CREATE = auto()
-        PRE_SCHEMA = auto()
-        POST_INDEX = auto()
+        POST_NAME = auto()
         POST_SCHEMA_ROOT = auto()
         POST_SCHEMA_WITH = auto()
```
Comment (tobymao) [inline sqlglot/expressions.py:1508]: this can just be called WITH

Comment (tobymao) [reply sqlglot/expressions.py:1508]: or to be consistent, POST_WITH is fine

File: sqlglot/parser.py [hunk 3]
```diff
@@ -1275,6 +1302,51 @@ def _parse_withisolatedloading(self) -> exp.Expression:
...
 
+    def _parse_locking(self) -> exp.Expression:
+        if self._match(TokenType.TABLE):
+            what = "TABLE"
+        elif self._match(TokenType.VIEW):
+            what = "VIEW"
+        elif self._match(TokenType.ROW):
+            what = "ROW"
+        elif self._match_text_seq("DATABASE"):
+            what = "DATABASE"
```
Comment (tobymao) [inline sqlglot/parser.py:1313]: syntax error, add an else what  = None or something

Comment (tobymao) [reply sqlglot/parser.py:1313]: instead of what, lets call it kind which is more consistent with create

File: sqlglot/parser.py [hunk 4]
```diff
@@ -1275,6 +1302,51 @@ def _parse_withisolatedloading(self) -> exp.Expression:
...
+        elif self._match_texts(("EXCL", "EXCLUSIVE")):
+            lock_type = "EXCLUSIVE"
+        elif self._match_text_seq("SHARE"):
+            lock_type = "SHARE"
+        elif self._match_text_seq("READ"):
+            lock_type = "READ"
+        elif self._match_text_seq("WRITE"):
+            lock_type = "WRITE"
+        elif self._match_text_seq("CHECKSUM"):
+            lock_type = "CHECKSUM"
```
Comment (tobymao) [inline sqlglot/parser.py:1335]: same, syntax error

File: sqlglot/parser.py [hunk 5]
```diff
@@ -1275,6 +1302,51 @@ def _parse_withisolatedloading(self) -> exp.Expression:
...
+
+        override = None
```
Comment (tobymao) [inline sqlglot/parser.py:1337]: override = self._match_text_seq("OVERRIDE")

File: sqlglot/generator.py [hunk 6]
```diff
@@ -910,6 +922,15 @@ def isolatedloadingproperty_sql(self, expression: exp.IsolatedLoadingProperty) -
             for_ = " FOR NONE"
         return f"WITH{no}{concurrent} ISOLATED LOADING{for_}"
 
+    def lockingproperty_sql(self, expression: exp.LockingProperty) -> str:
+        what = expression.args.get("what")
+        this = expression.args.get("this")
+        this = f" {this}" if this else ""
```
Comment (GeorgeSittas) [inline sqlglot/generator.py:928]:
```suggestion
        this = f" {this}" if expression.this else ""
```

File: sqlglot/expressions.py [hunk 7]
```diff
@@ -1454,6 +1454,16 @@ class IsolatedLoadingProperty(Property):
     }
 
 
+class LockingProperty(Property):
+    arg_types = {
+        "this": False,
+        "what": True,
```
Comment (tobymao) [inline sqlglot/expressions.py:1460]: change this to kind

File: sqlglot/expressions.py [hunk 8]
```diff
@@ -1479,12 +1489,25 @@ class Properties(Expression):
...
+    #   create [POST_CREATE]
+    #     table a [POST_NAME]
+    #     as [POST_ALIAS] (select * from b)
+    #     index (c) [POST_INDEX]
     class Location(AutoName):
         POST_CREATE = auto()
-        PRE_SCHEMA = auto()
-        POST_INDEX = auto()
+        POST_NAME = auto()
         POST_SCHEMA_ROOT = auto()
```
Comment (tobymao) [inline sqlglot/expressions.py:1507]: POST_SCHEMA

File: sqlglot/dialects/teradata.py (hunk 1 above)
Comment (treysp) [reply sqlglot/dialects/teradata.py:13]: No lie!

https://docs.teradata.com/r/Teradata-Studio-Express-User-Guide/June-2017/Query-Development/Teradata-SQL-Editor/SQL-Editor-Templates/Inserting-Aster-Analytical-Functions

image

**Review by tobymao — APPROVED** [review]

**Pull request merged**